core/sail/base/src/main/java/org/eclipse/rdf4j/sail/base/SailSourceConnection.java

```diff
@@ -1033,4 +1034,21 @@ private boolean hasStatement(SailDataset dataset, Resource subj, IRI pred, Value
 		}
 	}
 
+	/**
+	 * Returns the number of statements in the snapshot, optionally including inferred statements, for the given
+	 * contexts. This method reads the size directly from the dataset within the current isolation level.
+	 *
+	 * @param includeInferred whether to include inferred statements in the count
+	 * @param contexts        the RDF contexts (named graphs) to restrict the count to; if none are provided, counts all
+	 *                        contexts
+	 * @return the number of statements in the dataset
+	 * @throws SailException if an error occurs while accessing the Sail store
+	 */
+	@Experimental
+	protected long getSizeFromSnapshot(final boolean includeInferred, final Resource... contexts) throws SailException {
```

Comment:
Do we need `includeInferred`? I can't see that the other size related methods support differentiating between inferred or explicit.

Reply:
I kinda wish we could just use the existing sizeInternal(...) method, but I get that it's better to be safe and introduce a separate method for now and then later on we can consider changing up sizeInternal(...).